File: src/treemapper/diffctx/edges/semantic/clojure.py
from __future__ import annotations

import re
from collections import defaultdict
from pathlib import Path

from ...config.weights import EDGE_WEIGHTS
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict, FragmentIndex, discover_files_by_refs

_CLOJURE_EXTS = {".clj", ".cljs", ".cljc", ".edn"}

_CLOJURE_NS_REQUIRE_RE = re.compile(
    r"\(:?require\s[^)]{0,2000}?\[([a-z][\w.-]{0,200})",
)
_CLOJURE_REQUIRE_RE = re.compile(
    r"\(require\s{1,10}['\"]?([a-z][\w.-]{0,200})",
)

_DEFN_RE = re.compile(r"^\s*\(defn-?\s+([a-z_][\w*+!?<>=-]{0,100})", re.MULTILINE)
_DEF_RE = re.compile(r"^\s*\(def\s+([a-z_][\w*+!?<>=-]{0,100})", re.MULTILINE)
_DEFPROTOCOL_RE = re.compile(r"^\s*\(defprotocol\s+([A-Z]\w{0,100})", re.MULTILINE)
_DEFRECORD_RE = re.compile(r"^\s*\(defrecord\s+([A-Z]\w{0,100})", re.MULTILINE)
_DEFTYPE_RE = re.compile(r"^\s*\(deftype\s+([A-Z]\w{0,100})", re.MULTILINE)
_DEFMULTI_RE = re.compile(r"^\s*\(defmulti\s+([a-z_][\w*+!?<>=-]{0,100})", re.MULTILINE)
_EXTEND_PROTOCOL_RE = re.compile(r"\(extend-(?:protocol|type)\s+([A-Z]\w{0,100})", re.MULTILINE)
_IMPLEMENTS_RE = re.compile(r"\((?:reify|proxy)\s+\[?([A-Z]\w{0,100})", re.MULTILINE)

_FUNC_CALL_RE = re.compile(r"\(([a-z_][\w*+!?<>=-]{1,100})\b")
_TYPE_REF_RE = re.compile(r"(?<![a-z_])([A-Z]\w{1,100})\b")
_QUALIFIED_CALL_RE = re.compile(r"\(([a-z][\w.-]{0,100})/([a-z_][\w*+!?<>=-]{1,100})")

_CLOJURE_KEYWORDS = frozenset(
    {
        "if",
        "do",
        "let",
        "fn",
        "def",
        "defn",
        "defn-",
        "defmacro",
        "loop",
        "recur",
        "when",
        "cond",
        "case",
        "and",
        "or",
        "not",
        "throw",
        "try",
        "catch",
        "finally",
        "new",
        "set!",
        "quote",
        "var",
        "import",
        "require",
        "use",
        "ns",
        "in-ns",
        "refer",
        "println",
        "print",
        "str",
        "prn",
        "pr",
        "pr-str",
        "first",
        "rest",
        "cons",
        "conj",
        "assoc",
        "dissoc",
        "get",
        "nth",
        "count",
        "map",
        "filter",
        "reduce",
        "apply",
        "comp",
        "partial",
        "into",
        "seq",
        "vec",
        "list",
        "hash-map",
        "sorted-map",
        "atom",
        "deref",
        "swap!",
        "reset!",
        "compare-and-set!",
        "nil",
        "true",
        "false",
    }
)

_DIFF_REQUIRE_RE = re.compile(r"^\+.*\[:?require\s[^]]{0,2000}?\[([a-z][\w.-]{0,200})", re.MULTILINE)
_DIFF_REQUIRE_SIMPLE_RE = re.compile(r"^\+\s*\(require\s{1,10}['\"]?([a-z][\w.-]{0,200})", re.MULTILINE)


def _is_clojure_file(path: Path) -> bool:
    return path.suffix.lower() in _CLOJURE_EXTS


def _extract_requires(content: str) -> set[str]:
    refs: set[str] = set()
    for m in _CLOJURE_NS_REQUIRE_RE.finditer(content):
        refs.add(m.group(1))
    for m in _CLOJURE_REQUIRE_RE.finditer(content):
        refs.add(m.group(1))
    return refs


def _extract_definitions(content: str) -> tuple[set[str], set[str]]:
    funcs: set[str] = set()
    for m in _DEFN_RE.finditer(content):
        funcs.add(m.group(1))
    for m in _DEF_RE.finditer(content):
        funcs.add(m.group(1))
    for m in _DEFMULTI_RE.finditer(content):
        funcs.add(m.group(1))

    types: set[str] = set()
    types.update(m.group(1) for m in _DEFPROTOCOL_RE.finditer(content))
    types.update(m.group(1) for m in _DEFRECORD_RE.finditer(content))
    types.update(m.group(1) for m in _DEFTYPE_RE.finditer(content))

    return funcs, types


def _extract_protocol_refs(content: str) -> set[str]:
    refs: set[str] = set()
    for m in _EXTEND_PROTOCOL_RE.finditer(content):
        refs.add(m.group(1))
    for m in _IMPLEMENTS_RE.finditer(content):
        refs.add(m.group(1))
    return refs


def _extract_references(content: str) -> tuple[set[str], set[str], set[str]]:
    func_calls = {m.group(1) for m in _FUNC_CALL_RE.finditer(content) if m.group(1) not in _CLOJURE_KEYWORDS}

    type_refs = {m.group(1) for m in _TYPE_REF_RE.finditer(content)}

    qualified_ns: set[str] = set()
    for m in _QUALIFIED_CALL_RE.finditer(content):
        qualified_ns.add(m.group(1))

    return func_calls, type_refs, qualified_ns


def _namespace_to_path(ns: str) -> str:
    return ns.replace(".", "/").replace("-", "_")


def _namespace_leaf(ns: str) -> str:
    return ns.rsplit(".", maxsplit=1)[-1].lower().replace("-", "_")


def _collect_clojure_refs(clojure_files: list[Path]) -> set[str]:
    refs: set[str] = set()
    for cf in clojure_files:
        try:
            content = cf.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            continue
        for ns in _extract_requires(content):
            refs.add(_namespace_to_path(ns))
            refs.add(_namespace_leaf(ns))
    return refs


class ClojureEdgeBuilder(EdgeBuilder):
    weight = 0.60
    require_weight = EDGE_WEIGHTS["clojure_require"].forward
    fn_weight = EDGE_WEIGHTS["clojure_fn"].forward
    protocol_weight = EDGE_WEIGHTS["clojure_protocol"].forward
    reverse_weight_factor = EDGE_WEIGHTS["clojure_require"].reverse_factor

    def needs_from_diff(self, diff_content: str) -> list[str]:
        refs: list[str] = []
        for m in _DIFF_REQUIRE_RE.finditer(diff_content):
            refs.append(_namespace_to_path(m.group(1)))
            refs.append(_namespace_leaf(m.group(1)))
        for m in _DIFF_REQUIRE_SIMPLE_RE.finditer(diff_content):
            refs.append(_namespace_to_path(m.group(1)))
            refs.append(_namespace_leaf(m.group(1)))
        return refs

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
        **kwargs: object,
    ) -> list[Path]:
        clojure_files = [f for f in changed_files if _is_clojure_file(f)]
        if not clojure_files:
            return []

        refs = _collect_clojure_refs(clojure_files)
        return discover_files_by_refs(refs, changed_files, all_candidate_files, repo_root)

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        clojure_frags = [f for f in fragments if _is_clojure_file(f.path)]
        if not clojure_frags:
            return {}

        edges: EdgeDict = {}
        idx = FragmentIndex(fragments, repo_root)
        fn_defs, type_defs = self._build_indices(clojure_frags)

        for cf in clojure_frags:
            self._add_fragment_edges(cf, idx, fn_defs, type_defs, edges)

        return edges

    def _build_indices(self, clojure_frags: list[Fragment]) -> tuple[dict[str, list[FragmentId]], dict[str, list[FragmentId]]]:
        fn_defs: dict[str, list[FragmentId]] = defaultdict(list)
        type_defs: dict[str, list[FragmentId]] = defaultdict(list)

        for f in clojure_frags:
            funcs, types = _extract_definitions(f.content)
            for fn in funcs:
                fn_defs[fn.lower()].append(f.id)
            for t in types:
                type_defs[t.lower()].append(f.id)

        return fn_defs, type_defs

    def _link_requires(self, cf: Fragment, idx: FragmentIndex, edges: EdgeDict) -> None:
        for ns in _extract_requires(cf.content):
            self.link_by_stem(cf.id, _namespace_leaf(ns), idx, edges, self.require_weight)
            self.link_by_path_match(cf.id, _namespace_to_path(ns), idx, edges, self.require_weight)

    def _link_protocol_refs(
        self,
        cf: Fragment,
        type_defs: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for proto in _extract_protocol_refs(cf.content):
            for fid in type_defs.get(proto.lower(), []):
                if fid != cf.id:
                    self.add_edge(edges, cf.id, fid, self.protocol_weight)

    def _link_fn_calls(
        self,
        cf: Fragment,
        fn_defs: dict[str, list[FragmentId]],
        self_fn_lower: set[str],
        edges: EdgeDict,
    ) -> None:
        func_calls, _, _ = _extract_references(cf.content)
        for func_call in func_calls:
            if func_call.lower() in self_fn_lower:
                continue
            for fid in fn_defs.get(func_call.lower(), []):
                if fid != cf.id:
                    self.add_edge(edges, cf.id, fid, self.fn_weight)

    def _link_type_refs(
        self,
        cf: Fragment,
        type_defs: dict[str, list[FragmentId]],
        self_type_lower: set[str],
        edges: EdgeDict,
    ) -> None:
        _, type_refs, _ = _extract_references(cf.content)
        for type_ref in type_refs:
            if type_ref.lower() in self_type_lower:
                continue
            for fid in type_defs.get(type_ref.lower(), []):
                if fid != cf.id:
                    self.add_edge(edges, cf.id, fid, self.protocol_weight)

    def _link_qualified_ns(self, cf: Fragment, idx: FragmentIndex, edges: EdgeDict) -> None:
        _, _, qualified_ns = _extract_references(cf.content)
        for ns in qualified_ns:
            self.link_by_stem(cf.id, _namespace_leaf(ns), idx, edges, self.require_weight)

    def _add_fragment_edges(
        self,
        cf: Fragment,
        idx: FragmentIndex,
        fn_defs: dict[str, list[FragmentId]],
        type_defs: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        self._link_requires(cf, idx, edges)
        self._link_protocol_refs(cf, type_defs, edges)

        self_funcs, self_types = _extract_definitions(cf.content)
        self_fn_lower = {fn.lower() for fn in self_funcs}
        self_type_lower = {t.lower() for t in self_types}

        self._link_fn_calls(cf, fn_defs, self_fn_lower, edges)
        self._link_type_refs(cf, type_defs, self_type_lower, edges)
        self._link_qualified_ns(cf, idx, edges)

File: diffctx/src/edges/semantic/clojure.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::weights::EDGE_WEIGHTS;
use crate::types::Fragment;

use super::super::EdgeDict;
use super::super::base::{self, EdgeBuilder, add_edge, add_edges_from_ids, discover_files_by_refs};

fn is_clojure_file(path: &Path) -> bool {
    let ext = base::file_ext(path);
    matches!(ext.as_str(), ".clj" | ".cljs" | ".cljc" | ".edn")
}

static NS_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\(ns\s+([\w.\-]+)").unwrap());
static REQUIRE_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r":require\s*\[([^\]]+)\]").unwrap());
static REQUIRE_SINGLE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\[?([\w.\-]+)(?:\s+:as\s+(\w+))?").unwrap());
static USE_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r":use\s*\[([^\]]+)\]").unwrap());
static IMPORT_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r":import\s*\[([^\]]+)\]").unwrap());
static DEFN_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?m)^\s*\((?:defn-?|def|defmacro|defprotocol|defrecord|deftype|defmulti|defmethod|defonce)\s+([\w\-!?*+<>=]+)").unwrap()
});
static CALL_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\(([\w.\-]+/[\w\-!?*+<>=]+)").unwrap());

static CLJ_KEYWORDS: Lazy<FxHashSet<&str>> = Lazy::new(|| {
    [
        "if", "do", "let", "fn", "def", "defn", "defmacro", "when", "cond", "case", "loop",
        "recur", "throw", "try", "catch", "finally", "quote", "var", "ns", "require", "use",
        "import", "in-ns", "refer", "nil", "true", "false", "println", "pr", "prn", "str", "first",
        "rest", "cons", "conj", "assoc", "dissoc", "get", "count", "map", "filter", "reduce",
        "apply", "partial", "comp", "identity", "not",
    ]
    .iter()
    .copied()
    .collect()
});

fn extract_requires(content: &str) -> FxHashSet<String> {
    let mut refs = FxHashSet::default();
    for cap in REQUIRE_RE.captures_iter(content) {
        for single in REQUIRE_SINGLE_RE.captures_iter(&cap[1]) {
            let name = &single[1];
            if !name.starts_with(':') {
                refs.insert(name.to_string());
                if let Some(leaf) = name.split('.').last() {
                    refs.insert(leaf.to_string());
                }
            }
        }
    }
    for cap in USE_RE.captures_iter(content) {
        for single in REQUIRE_SINGLE_RE.captures_iter(&cap[1]) {
            refs.insert(single[1].to_string());
        }
    }
    for cap in IMPORT_RE.captures_iter(content) {
        for word in cap[1].split_whitespace() {
            let w = word.trim_matches(|c: char| !c.is_alphanumeric() && c != '.');
            if !w.is_empty() {
                refs.insert(w.to_string());
            }
        }
    }
    refs
}

fn extract_ns(content: &str) -> Option<String> {
    NS_RE.captures(content).map(|c| c[1].to_string())
}

fn extract_defs(content: &str) -> FxHashSet<String> {
    DEFN_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .collect()
}

pub struct ClojureEdgeBuilder;

impl EdgeBuilder for ClojureEdgeBuilder {
    fn build(&self, fragments: &[Fragment], repo_root: Option<&Path>) -> EdgeDict {
        let frags: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_clojure_file(Path::new(f.path())))
            .collect();
        if frags.is_empty() {
            return FxHashMap::default();
        }

        let require_w = EDGE_WEIGHTS["clojure_require"].forward;
        let fn_w = EDGE_WEIGHTS["clojure_fn"].forward;
        let _proto_w = EDGE_WEIGHTS["clojure_protocol"].forward;
        let reverse_factor = EDGE_WEIGHTS["clojure_require"].reverse_factor;

        let idx = base::FragmentIndex::new(fragments, repo_root);
        let mut ns_to_frags: FxHashMap<String, Vec<_>> = FxHashMap::default();
        let mut name_to_defs: FxHashMap<String, Vec<_>> = FxHashMap::default();
        for f in &frags {
            if let Some(ns) = extract_ns(&f.content) {
                let leaf = ns.split('.').last().unwrap_or(&ns).to_lowercase();
                ns_to_frags.entry(leaf).or_default().push(f.id.clone());
                ns_to_frags
                    .entry(ns.to_lowercase())
                    .or_default()
                    .push(f.id.clone());
            }
            for name in extract_defs(&f.content) {
                name_to_defs
                    .entry(name.to_lowercase())
                    .or_default()
                    .push(f.id.clone());
            }
        }

        let mut edges: EdgeDict = FxHashMap::default();

        for f in &frags {
            let self_defs = extract_defs(&f.content);
            for req in extract_requires(&f.content) {
                let leaf = req.split('.').last().unwrap_or(&req).to_lowercase();
                if let Some(targets) = ns_to_frags.get(&leaf) {
                    add_edges_from_ids(&mut edges, &f.id, targets, require_w, reverse_factor);
                } else {
                    base::link_by_name(&f.id, &req, &idx, &mut edges, require_w, reverse_factor);
                }
            }
            for cap in CALL_RE.captures_iter(&f.content) {
                let full = &cap[1];
                if let Some(func) = full.split('/').last() {
                    if self_defs.contains(func) || CLJ_KEYWORDS.contains(func) {
                        continue;
                    }
                    if let Some(targets) = name_to_defs.get(&func.to_lowercase()) {
                        for t in targets {
                            if t != &f.id {
                                add_edge(&mut edges, &f.id, t, fn_w, reverse_factor);
                            }
                        }
                    }
                }
            }
            for id in &f.identifiers {
                if self_defs.contains(id) || CLJ_KEYWORDS.contains(id.as_str()) {
                    continue;
                }
                if let Some(targets) = name_to_defs.get(&id.to_lowercase()) {
                    for t in targets {
                        if t != &f.id {
                            add_edge(&mut edges, &f.id, t, fn_w, reverse_factor);
                        }
                    }
                }
            }
        }
        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let clj_changed: Vec<&PathBuf> = changed.iter().filter(|f| is_clojure_file(f)).collect();
        if clj_changed.is_empty() {
            return vec![];
        }
        let mut refs = FxHashSet::default();
        for f in &clj_changed {
            if let Some(content) = base::read_file_cached(f, file_cache) {
                refs.extend(extract_requires(&content));
            }
        }
        discover_files_by_refs(&refs, changed, candidates, repo_root)
    }
}
